Add documentation about migrate with horizontal migration (#1401)
PrestaShop/docs: development/architecture/migration-guide/horizontal-migration/migrate-list.md

---
title: Cleaning up
weight: 60
---

# Migrate list

> Once you create your controller,and you make your routing file. We can start to have fun with migration.

Reviewer: ```suggestion
Once you have your controller and routing file prepared, you can start migrating other parts. It is quite convenient to start with the list as it is usually the main part of the index page.
```

## Routing

> We will start with the list. You have to add to the routing file an entry for the list.

Reviewer: ```suggestion
Configure a route for index action in your routing file. (See more about symfony routing in https://symfony.com/doc/4.4/routing.html)
```

> Keep our example with TitleController.

Reviewer: ```suggestion
```

Example:
```yaml
admin_title_index:
  path: /
  methods: [ GET, POST ]
  defaults:
    _controller: PrestaShopBundle\Controller\Admin\Configure\ShopParameters\TitleController::indexAction
    _legacy_controller: AdminGender
    _legacy_link: AdminGender
```

> You must specify as

Reviewer: ```suggestion
As you can see from example - you must specify some default parameters
```

- `_controller` the namespace of your controller
- `_legacy_controller` the name of the legacy controller

> - `_legacy_link` the name of the legacy action (for the list it’s the same as the legacy controller)

Reviewer: maybe it is worth briefly describing what are every one of these params used for and maybe provide some indication of what could go wrong (like some errors examples if some of them are missing)

> This information will be helpful to redirect the legacy link to your controller automatically

Reviewer: if above comment is taken into account then this part may become redundant.

## Controller

> Once you have finished your routing, your controller must implements this interface: `PrestaShopBundle\Bridge\AdminController\LegacyListControllerBridgeInterface`.

Reviewer: ```suggestion
Once the routing is complete, we can proceed with the controller.
Every contoller containing a list, must implement the `PrestaShopBundle\Bridge\AdminController\LegacyListControllerBridgeInterface`. This interface requires to implement these methods:
* `getPositionIdentifier`: have to return position identifier for legacy list request
* `getIdentifier`: have to return the identifier for related database table
```

> The previous interface needs these two method:

Reviewer: The above-mentioned interface requires to implement these methods:

> The previous interface needs these two method:
> * `getPositionIdentifier`: have to return position identifier for legacy list request
> * `getIdentifier`: have to return the identifier for table

Reviewer: ```suggestion
```

Then you have to create the corresponding action.

```php

> public function index(Request $request)

Reviewer: ```suggestion
public function index(Request $request): Response
```

{
    ....
}
```

> **Get an instance of `HelperListConfiguration`**

Reviewer: Missing some abstract info about the bridges before jumping straight to one of them.
Something like:
`Now its time to start creating the "bridges" (link to related article maybe), which will take care of symfony adaptation to internal legacy services.` (you can probably find better wording, but you get the idea)

- The name of the table like in legacy (for our example, it's `gender`)
- The name of the object model class in legacy (for our example, it's `Gender`)
- An instance of controller configuration, which is an object

### Define the list of fields
To add a new field to the list, your controller must use the `addListField` methods from the `AdminControllerTrait`.

This method takes an instance of `FieldInterface` and an instance of `HelperListConfiguration`.And she adds the field to the field list in HelperListConfiguration.

A `FieldInterface` object needs a label and an array of configurations. In this array of configurations, you can define the following options:

- title: which is required and represents the text, will be shown in the list's header
- align: represents where you want to align the value in the cell
- class: represents a class which be added to the class html attributes in the cell
- filter_key: represents the name of the filter key
- orderby: tell if you want or not to be able to order by this field in your list
- position: identify a field as a position cell
- search: enable or disable the search by this field
- width: define the width for the cell of this field

> You can find all the configuration of your field in the legacy controller. You have to copy/paste it at the second `config` attribute of the field class.

Reviewer: define where can be the example of titleController found in legacy

> > Example for the hypothetic TitleController and for the field `id_gender`:

Reviewer: ```suggestion
> Example for the hypothetic TitleController for the field `id_gender`:
```

>

```php
$this->addListField(new Field(
        'id_gender', [
        'title' => $this->trans('ID', [], 'Admin.Global'),
        'align' => 'center',
        'class' => 'fixed-width-xs',
    ]
), $helperListConfiguration);
```

### Define actions to your user
You have two way to add actions, these two methods are in the `AdminControllerTrait`:

- `addAction` which handle `HeaderToolbarAction`
- `addActionList` which handle `ListHeaderToolbarAction`, `ListRowAction`, `ListBulkAction`

The `HeaderToolbarAction` represents an action in the header toolbar of the page.

The `ListHeaderToolbarAction` represents an action in the header toolbar of the list.

The `ListRowAction` represents an action in each row of the list.

The `ListBulkAction` represents a bulk action for the list.

To create an user action, you must define it as a label and an array of configurations. In this configuration, you can specify the following options:

- confirm: is the text for the confirmation alert
- desc: is the text for the title HTML attribute
- href: the link to the action
- icon: a name of a PrestaShop icons

> - text: the text in the link for the action

Reviewer: I would suggest first listing what actions are available (would be nice providing images of where they are). And maybe then mention methods in AdminControllerTrait (or mention related methods next to related action like: 
```The `ListRowAction` (image here) represents an action in each row of the list. You can add it via addActionList method in AdminControllerTrait)```.
Or maybe this even deserves a table defining columns, actions, example images and methods (not sure :man_shrugging: )
